SwiftGit2/Repository.swift

```diff
@@ -537,4 +537,161 @@ final public class Repository {
 		let iterator = CommitIterator(repo: self, root: branch.oid.oid)
 		return iterator
 	}
+
+	// MARK: - Diffs
+
+	public func diff(for commit: Commit) -> Result<[Diff.Delta], NSError> {
+		/// Get the Base Tree
+		let baseCommit = self.commit(with: commit.oid)
+		guard baseCommit.error == nil else {
+			return Result.failure(baseCommit.error!)
+		}
+
+		let baseTree = self.tree(from: baseCommit.value!)
+		guard baseTree.error == nil else {
+			return Result.failure(baseTree.error!)
+		}
```

Comment:
No that these are nicely abstracted, we can use `flatMap` to propagate errors automatically.

```swift
return self
    .flatMap { baseCommit in
        return self.tree(from: baseCommit)
    }
    .flatMap { baseTree in
        guard !commit.parents.isEmpty else {
            // Initial commit in a repository
            return self.diff(withOldTree: nil, andNewTree: baseTree.value)
        }

        var mergeDiff: Result<OpaquePointer?, NSError> = .success(nil)
        for parent in commit.parents {
            mergeDiff = mergeDiff
                .flatMap { mergeDiff in
                    return self
                        .parentCommit(from: parent)
                        .flatMap { commit in
                            return self.tree(from: commit)
                        }
                        .flatMap { tree in
                            return self.diff(withOldTree: tree, andNewTree: baseTree)
                        }
                        .flatMap { diff in
                            guard let mergeDiff = mergeDiff else { return .success(diff) }
                            let merge = git_diff_merge(mergeDiff, diff)
                            guard mergeResult == gIT_OK.rawValue else {
                                // return return error
                            }
                            return .success(merge)
                        }
                }
        }
    }
    .flatMap { diffResult in
        return self.processDiffDeltas(diffResult)
    }
```

Reply:
interesting. I've seen flatmap before but not like this. I'll put this in

Reply:
Wait, is flatMap supposed to be implemented on Repository? I see it implemented in Result but I'm not sure how to use this sample code.

Reply:
Oops! There was a line missing. I've updated the code sample above.